src/main/haskell/kore/src/Kore/Step/Substitution.hs

```diff
@@ -292,3 +303,96 @@ normalizePredicatedSubstitution
             )
         Right (Predicated { predicate = p, substitution = s }, _) ->
             (Predicated term p s, EmptyUnificationProof)
+
+{-| Creates a 'PredicateSubstitutionMerger' that returns errors on unifications it
+can't handle.
+-}
+createPredicatesAndSubstitutionsMergerExcept
+    :: forall level variable m err .
+        ( Show (variable level)
+        , Unparse (variable level)
+        , SortedVariable variable
+        , MetaOrObject level
+        , Ord (variable level)
+        , OrdMetaOrObject variable
+        , ShowMetaOrObject variable
+        , FreshVariable variable
+        , Monad m
+        , err ~ ExceptT (UnificationOrSubstitutionError level variable)
+        )
+    => MetadataTools level StepperAttributes
+    -> PredicateSubstitutionSimplifier level m
+    -> PredicateSubstitutionMerger level variable (err m)
+createPredicatesAndSubstitutionsMergerExcept tools substitutionSimplifier =
+    PredicateSubstitutionMerger worker
+  where
+    worker
+        :: [Predicate level variable]
+        -> [Substitution level variable]
+        -> (err m) (PredicateSubstitution level variable)
+    worker predicates substitutions = do
+        (merged, _proof) <- mergePredicatesAndSubstitutionsExcept
+            tools substitutionSimplifier predicates substitutions
+        return merged
+
+{-| Creates a 'PredicateSubstitutionMerger' that creates predicates for
+unifications it can't handle.
+-}
+createPredicatesAndSubstitutionsMerger
+    :: forall level variable m .
+        ( Show (variable level)
+        , Unparse (variable level)
+        , SortedVariable variable
+        , MetaOrObject level
+        , Ord (variable level)
+        , OrdMetaOrObject variable
+        , ShowMetaOrObject variable
+        , FreshVariable variable
+        , Monad m
+        )
+    => MetadataTools level StepperAttributes
+    -> PredicateSubstitutionSimplifier level m
+    -> PredicateSubstitutionMerger level variable m
+createPredicatesAndSubstitutionsMerger tools substitutionSimplifier =
+    PredicateSubstitutionMerger worker
+  where
+    worker
+        :: [Predicate level variable]
+        -> [Substitution level variable]
+        -> m (PredicateSubstitution level variable)
+    worker predicates substitutions = do
+        (merged, _proof) <- mergePredicatesAndSubstitutions
```

Comment:
Does this do anything extra, beyond combining predicates with `\and` and normalizing the merged substitution? Do we need to normalize the substitution at this point?

Reply:
This is exactly what this does. This is called at the end of unification/matching, which, I think, should return a normalized substitution, but we could argue about this if you disagree.

Reply:
> This is called at the end of unification/matching, which, I think, should return a normalized substitution

It wasn't clear to me that this is only called at the end of unification and matching. In that case, I agree.